### PR TITLE
feat: [PL-23377]: SCM, Azure createhook for multiple events

### PR DESCRIPTION
**Pull request by tphoney**
You can use the following comments to re-trigger PR Checks

- Compile: `trigger compile`
- runAeriformCheck: `trigger AeriformCheck`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`

---
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/32016)

### CODE REVIEW

**Review by mohitg0795 — DISMISSED** [review]
:lgtm:

> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/32016#-:-MzqA96f2dncrIIKwOnE:bnfp4nl)* status: 3 file(s) to review. 1 of 2 approvals obtained.

File: product/ci/scm/proto/scm.proto [hunk 1]
```diff
@@ -602,8 +602,7 @@ message BitbucketServerWebhookEvents {
 }
 
 message AzureWebhookEvents {
-  // Azure only allows one event type per hook
-  AzureWebhookEvent events = 1;
+  repeated AzureWebhookEvent events = 1;
```
Comment (shubham149) [inline product/ci/scm/proto/scm.proto:605]: This is not backward compatible change.

File: product/ci/scm/repo/repo.go [hunk 2]
```diff
@@ -328,35 +355,37 @@ func convertBitbucketCloudEnumToStrings(enums *pb.BitbucketCloudWebhookEvents) (
...
+		case pb.AzureWebhookEvent_AZURE_PULLREQUEST_MERGED:
+			strings = append(strings, "git.pullrequest.merged")
+		}
 	}
 	return strings
 }
 
 func convertStringsToAzureEnum(strings []string) (enums pb.NativeEvents_Azure) {
-	var array []pb.AzureWebhookEvent
+	var sliceOfEvents []pb.AzureWebhookEvent
```
Comment (shubham149) [inline product/ci/scm/repo/repo.go:374]: Use events instead of sliceOfEvents

File: product/ci/scm/repo/repo.go [hunk 3]
```diff
@@ -65,23 +65,54 @@ func CreateWebhook(ctx context.Context, request *pb.CreateWebhookRequest, log *z
...
-			"elapsed_time_ms", utils.TimeSince(start), zap.Error(err))
-		// this is a hard error with no response
-		if response == nil {
-			return nil, err
+	var hook *scm.Hook
+	var response *scm.Response
+	switch client.Driver {
+	case scm.DriverAzure:
+		// iterate through the events and create a hook for each one
+		var createdEvents []string
```
Comment (shubham149) [inline product/ci/scm/repo/repo.go:73]: Move the functionality of azure driver to a separate method. Current function has become too big.

Comment (tphoney) [reply product/ci/scm/repo/repo.go:73]: i had initially done this, but it led to more duplicated code, and an unwieldy method signature.
you need to pass through, ctx context.Context, request *pb.CreateWebhookRequest, log *zap.SugaredLogger, request, events, hook, inputparams.

this is a major work around for azure, and by its very nature messy.

File: product/ci/scm/repo/repo.go (hunk 2 above)
Comment (tphoney) [reply product/ci/scm/repo/repo.go]: updated.

**Review by tphoney — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/32016)* status: 3 file(s) to review. 3. unresolved discussions. 0 of 2 approvals obtained. 

___
*[product/ci/scm/proto/scm.proto, line 605 at r1](https://reviewable.io/reviews/harness/harness-core/32016#-Mzp5R5iD_KBn6vehy78-r1-605:-Mztn0xb4ywian6cxqG5:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/eb9b89aaa9a0f95a538115a1b4b56d7badabe62b/product/ci/scm/proto/scm.proto#L605)):*
<details><summary><i>Previously, shubham149 (Shubham Agrawal) wrote…</i></summary><blockquote>

This is not backward compatible change.
</blockquote></details>

Done.

File: product/ci/scm/repo/repo.go (hunk 3 above)
Comment (tphoney) [reply product/ci/scm/repo/repo.go:73]: I moved what i could into a function, 
when errors occur, we have to return them, so it does not suit moving more logic out into functions.

File: product/ci/scm/repo/repo.go [hunk 4]
```diff
@@ -103,6 +124,23 @@ func CreateWebhook(ctx context.Context, request *pb.CreateWebhookRequest, log *z
 	return out, nil
 }
 
+// processCreateHookResponse, decides wether to fail hard, or return the http status
+func processCreateHookResponse(response *scm.Response, err error, log *zap.SugaredLogger, request *pb.CreateWebhookRequest, start time.Time) (returnVal *pb.CreateWebhookResponse) {
```
Comment (shubham149) [inline product/ci/scm/repo/repo.go:128]: This is not a good approach as per me. Error handling needs to be in the error origin method.

File: product/ci/scm/repo/repo.go [hunk 5]
```diff
@@ -65,23 +65,44 @@ func CreateWebhook(ctx context.Context, request *pb.CreateWebhookRequest, log *z
...
-			return nil, err
+	var hook *scm.Hook
+	var response *scm.Response
+	switch client.Driver {
+	case scm.DriverAzure:
+		// iterate through the events and create a hook for each one
+		var createdEvents []string
+		for _, event := range inputParams.NativeEvents {
+			// create webhook with a single event
+			tmpInputParams := inputParams
```
Comment (shubham149) [inline product/ci/scm/repo/repo.go:76]: `
func createAzureWebhook(ctx context.Context, slug, event string, params scm.HookInput) {
   params.NativeEvents = []string{event}
   hook, response, err = client.Repositories.CreateHook(ctx, slug, &params)
   if err != nil {
		log.Errorw("CreateWebhook failure", "provider", gitclient.GetProvider(*request.GetProvider()), "slug", request.GetSlug(), "name", request.GetName(), "target", request.GetTarget(),
		...
	}
   }

}
`

File: product/ci/scm/repo/repo.go (hunk 3 above)
Comment (shubham149) [reply product/ci/scm/repo/repo.go:73]: func createAzureWebhook(ctx, params, slug, event string) {
  
}

File: product/ci/scm/repo/repo.go (hunk 5 above)
Comment (shubham149) [inline product/ci/scm/repo/repo.go:76]: response, err := createAzureWebhook(ctx, parmas, slug, event, logr)
if err != nil {

}

File: product/ci/scm/repo/repo.go [hunk 6]
```diff
@@ -65,23 +65,44 @@ func CreateWebhook(ctx context.Context, request *pb.CreateWebhookRequest, log *z
...
+			tmpInputParams.NativeEvents = []string{event}
+			hook, response, err = client.Repositories.CreateHook(ctx, request.GetSlug(), &tmpInputParams)
```
Comment (shubham149) [inline product/ci/scm/repo/repo.go:78]: Hook is getting replaced in this method. response needs to be updated as well here

**Comment by tphoney** [comment]
@mohitg0795 after discussing this change with @shubham149 , he was not comfortable with the implementation and suggests that we move this logic for creating multiple webhooks out of the SCM service and into the calling code.

**Pull request closed without merging**